boot/bootutil/zephyr/src/boot_request_retention.c

```diff
@@ -0,0 +1,333 @@
+
+/** Number of requests per image. */
+#define BOOT_REQUEST_PER_IMAGE 2
+
+/** Maximum number of request slots. */
+#define BOOT_REQUEST_SLOT_MAX_N (BOOT_REQUEST_PER_IMAGE * CONFIG_UPDATEABLE_IMAGE_NUMBER + 1)
+
+MCUBOOT_LOG_MODULE_REGISTER(bootloader_request);
+
+static const struct device *bootloader_request_dev =
+    DEVICE_DT_GET(DT_CHOSEN(zephyr_bootloader_request));
+BUILD_ASSERT((BOOT_REQUEST_SLOT_MAX_N * sizeof(uint8_t)) <
+             DT_REG_SIZE_BY_IDX(DT_CHOSEN(zephyr_bootloader_request), 0),
+             "zephyr,bootloader-request area is too small for bootloader request struct");
+
+enum boot_request_type {
+    /** Invalid request. */
+    BOOT_REQUEST_INVALID = 0,
+
+    /** Request a change in the bootloader boot mode.
+     *
+     * @details Use @p boot_request_mode as argument.
+     *          @p BOOT_REQUEST_IMG_BOOTLOADER as image number.
+     *
+     * @note Used to trigger recovery through i.e. retention sybsystem.
+     */
+    BOOT_REQUEST_BOOT_MODE = 1,
+
+    /** Select the preferred image to be selected during boot or update.
+     *
+     * @details Use @p boot_request_slot_t as argument.
+     *
+     * @note Used in the Direct XIP mode.
+     */
+    BOOT_REQUEST_IMG_PREFERENCE = 2,
+
+    /** Request a confirmation of an image.
+     *
+     * @details Use @p boot_request_slot_t as argument.
+     *
+     * @note Used if the code cannot modify the image trailer directly.
+     */
+    BOOT_REQUEST_IMG_CONFIRM = 3,
+};
+
+enum boot_request_slot {
+    /** Unsupported value. */
+    BOOT_REQUEST_SLOT_INVALID = 0,
+    /** Primary slot. */
+    BOOT_REQUEST_SLOT_PRIMARY = 1,
+    /** Secondary slot. */
+    BOOT_REQUEST_SLOT_SECONDARY = 2,
+};
+
+/** Alias type for the image and number. */
+typedef uint8_t boot_request_slot_t;
+
+enum boot_request_mode {
+    /** Execute a regular boot logic. */
+    BOOT_REQUEST_MODE_REGULAR = 0,
+    /** Execute the recovery boot logic. */
+    BOOT_REQUEST_MODE_RECOVERY = 1,
+    /** Execute the firmware loader logic. */
+    BOOT_REQUEST_MODE_FIRMWARE_LOADER = 2,
+    /** Unsupported value. */
+    BOOT_REQUEST_MODE_INVALID = 0xFF,
+};
+
+/** Alias type for the image number. */
+typedef uint8_t boot_request_img_t;
+
+/**
+ * @brief Find a variable slot for a given request.
+ *
+ * @param[in]  type   Type of request.
+ * @param[in]  image  Image number. Use @p BOOT_REQUEST_IMG_BOOTLOADER for generic requests.
+ * @param[out] slot   Variable slot to use.
+ *
+ * @return 0 on success; nonzero on failure.
+ */
+static int
+boot_request_slot_find(enum boot_request_type type, boot_request_img_t image, size_t *slot)
+{
+    if (slot == NULL) {
+        return -EINVAL;
+    }
+
+    switch (type) {
+    case BOOT_REQUEST_BOOT_MODE:
+        *slot = 0;
+        break;
+    case BOOT_REQUEST_IMG_PREFERENCE:
+        *slot = 1 + image * BOOT_REQUEST_PER_IMAGE;
+        break;
+    case BOOT_REQUEST_IMG_CONFIRM:
+        *slot = 2 + image * BOOT_REQUEST_PER_IMAGE;
```

Comment:
use enum for what these values mean, can actually then have a `_COUNT` in the enum and use that for the number of states instead of just defining to 2 separately